chain-core/src/property.rs

```diff
@@ -0,0 +1,228 @@
+    fn diff_transaction(&self, transaction: &T) -> Result<Self::Update, Self::Error>;
+
+    /// create a combined Update from the given transactions
+    fn diff<'a, I>(&self, transactions: I) -> Result<Self::Update, Self::Error>
+    where
+        I: IntoIterator<Item = &'a T> + Sized,
+        T: 'a;
+
+    /// apply an update to the leger.
+    fn apply(&mut self, update: Self::Update) -> Result<&mut Self, Self::Error>;
+
+    /// this function is a helper that calls `diff` and `apply` methods to modify
+    /// the state of the Ledger.
+    fn update<'a, I>(&mut self, transactions: I) -> Result<&mut Self, Self::Error>
+    where
+        I: IntoIterator<Item = &'a T> + Sized,
+        T: 'a,
+    {
+        let update = self.diff(transactions)?;
+        self.apply(update)
+    }
+}
+
+/// interface for the leader selection algorithm
+///
+/// this is the interface that is responsible to verify the Block are
+/// created by the right Leaders (i.e. that everyone follows the
+/// consensus algorithm).
+///
+/// This is also the same interface that is used to detect if we are the
+/// leader for the block at the given date.
+pub trait LeaderSelection {
+    /// a leader selection Update. This is an atomic representation of
+    /// the set of changes to apply to the leader selection state.
+    ///
+    /// Having an atomic representation of the changes allow other
+    /// interesting properties:
+    ///
+    /// * generic testing;
+    /// * diff based storage;
+    ///
+    type Update;
+
+    /// the block that we will get the information from
+    type Block: Block;
+
+    /// Leader Selection error type
+    type Error: std::error::Error;
+
+    /// given a Block, create an Update diff to see what are the changes
+    /// that will come with this new block.
+    ///
+    /// This function is also responsible to check the validity of the block
+    /// within the blockchain but not to check the Transactional entities.
+    /// The transaction part are verified with the [`Transaction::diff`]
+    /// method.
+    ///
+    /// Here we want to check the validity of the consensus and of the block
+    /// signature.
+    fn diff(&self, input: &Self::Block) -> Result<Self::Update, Self::Error>;
+
+    /// apply the Update to the LeaderSelection
+    fn apply(&mut self, update: Self::Update) -> Result<(), Self::Error>;
+
+    /// return if this instance of the LeaderSelection is leader of the
+    /// blockchain at the given date.
+    ///
+    fn is_leader_at(&self, date: <Self::Block as Block>::Date) -> Result<bool, Self::Error>;
+}
+
+/// Define that an object can be written in a `Write` object or read from the
+/// `Read` object.
+pub trait Serializable: Sized {
+    type Error: std::error::Error + From<std::io::Error>;
+
+    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error>;
+
+    fn deserialize<R: std::io::Read>(reader: R) -> Result<Self, Self::Error>;
+}
+
+#[cfg(test)]
+pub mod test {
```

Comment:
I would rather name the public quickcheck API module something different, and reserve the `test` module for unit tests, as per convention.